README.md
# (MODX)EvolutionCMS.snippets.ddGetChunk

Snippet gets the chunk contents by its name. For example, it useful to get chunks inside JS code.

Also it can:
* Pass additional data to chunk for parsing.
* Escape special chars for JS.
* Remove empty placeholders.


## Requires

* PHP >= 5.6
* [(MODX)EvolutionCMS](https://github.com/evolution-cms/evolution) >= 1.1
* [(MODX)EvolutionCMS.libraries.ddTools](https://code.divandesign.ru/modx/ddtools) >= 0.60


## Installation


### Using [(MODX)EvolutionCMS.libraries.ddInstaller](https://github.com/DivanDesign/EvolutionCMS.libraries.ddInstaller)

Just run the following PHP code in your sources or [Console](https://github.com/vanchelo/MODX-Evolution-Ajax-Console):

```php
//Include (MODX)EvolutionCMS.libraries.ddInstaller
require_once(
	$modx->getConfig('base_path') .
	'assets/libs/ddInstaller/require.php'
);

//Install (MODX)EvolutionCMS.snippets.ddGetChunk
\DDInstaller::install([
	'url' => 'https://github.com/DivanDesign/EvolutionCMS.snippets.ddGetChunk',
	'type' => 'snippet'
]);
```

* If `ddGetChunk` is not exist on your site, `ddInstaller` will just install it.
* If `ddGetChunk` is already exist on your site, `ddInstaller` will check it version and update it if needed.


### Manually


#### 1. Elements → Snippets: Create a new snippet with the following data

1. Snippet name: `ddGetChunk`.
2. Description: `<b>2.4</b> Snippet gets the chunk contents by its name. For example, it useful to get chunks inside JS code.`.
3. Category: `Core`.
4. Parse DocBlock: `no`.
5. Snippet code (php): Insert content of the `ddGetChunk_snippet.php` file from the archive.


#### 2. Elements → Manage Files

1. Create a new folder `assets/snippets/ddGetChunk/`.
2. Extract the archive to the folder (except `ddGetChunk_snippet.php`).


## Parameters description

* `name`
	* Desctription: Chunk name or code via `@CODE:` prefix.
	* Valid values:
		* `stringChunkName`
		* `string` — use inline templates starting with `@CODE:`
	* **Required**
	
* `placeholders`
	* Desctription:
		Additional data has to be passed into the chunk.  
		Arrays are supported too: `some[a]=one&some[b]=two` => `[+some.a+]`, `[+some.b+]`; `some[]=one&some[]=two` => `[+some.0+]`, `[some.1]`.
	* Valid values:
		* `stringJsonObject` — as [JSON](https://en.wikipedia.org/wiki/JSON)
		* `stringHjsonObject` — as [HJSON](https://hjson.github.io/)
		* `stringQueryFormatted` — as [Query string](https://en.wikipedia.org/wiki/Query_string)
		* It can also be set as native PHP object or array (e. g. for calls through `\DDTools\Snippet::runSnippet` or `$modx->runSnippet`):
			* `arrayAssociative`
			* `object`
	* Default value: —
	
* `removeEmptyPlaceholders`
	* Desctription: Placeholders which have not values to be replaced by will be deleted from parsed chunk if the parameter equals 1.
	* Valid values:
		* `0`
		* `1`
	* Default value: `0`
	
* `escapeResultForJS`
	* Desctription: Escaping special chars for JS.
	* Valid values:
		* `0`
		* `1`
	* Default value: `0`


## Examples

_All examples are written using [HJSON](https://hjson.github.io/) for the `placeholders` parameter, but if you want you can use vanilla JSON instead._


### Get content of the form into JS code

```html
<script>
	var form = '[[ddGetChunk? &name=`someForm` &escapeResultForJS=`1`]]';
	
	//Insert form to page
	$('body').append(form);
</script>
```


### Get chunk with Ditto and send necessary additional data into it

```
[[ddGetChunk?
	&name=`someChunk`
	&placeholders=`{
		id: 33
		orderBy: "someTv ASC, pub_date DESC"
	}`
]]
```

Code of `someChunk`:

```html
<div class="some">
	<div class="someDesignDiv"></div>
	<div>
		[[Ditto?
			&startID=`[+id+]`
			&orderBy=`[+orderBy+]`
			&tpl=`someChunk_item`
		]]
	</div>
</div>
```


### Run the snippet through `\DDTools\Snippet::runSnippet` without DB and eval

```php
\DDTools\Snippet::runSnippet([
	'name' => 'ddGetChunk',
	'params' => [
		'name' => 'someChunk'
	]
]);
```


## Links

* [Home page](https://code.divandesign.ru/modx/ddgetchunk)
* [Telegram chat](https://t.me/dd_code)
* [Packagist](https://packagist.org/packages/dd/evolutioncms-snippets-ddgetchunk)
* [GitHub](https://github.com/DivanDesign/EvolutionCMS.snippets.ddGetChunk)


<link rel="stylesheet" type="text/css" href="https://raw.githack.com/DivanDesign/CSS.ddMarkdown/master/style.min.css" />

CHANGELOG.md
# (MODX)EvolutionCMS.snippets.ddGetChunk changelog


## Version 2.4 (2023-05-14)
* \+ Parameters → `placeholders`: Can also be set as a native PHP object/array.
* \* `\ddTools::getTpl` is used instead of `$modx->getTpl` (means a bit less bugs).
* \* Attention! (MODX)EvolutionCMS.libraries.ddTools >= 0.60 is required.


## Version 2.3 (2021-03-24)
* \* Attention! PHP >= 5.6 is required.
* \* Attention! (MODX)EvolutionCMS.libraries.ddTools >= 0.47 is required.
* \+ You can just call `\DDTools\Snippet::runSnippet` to run the snippet without DB and eval.
* \+ Parameters → `placeholders`: [HJSON](https://hjson.github.io/) is supported.
* \+ `\ddGetChunk\Snippet`: The new class. All snippet code moved here.


## Version 2.2.2 (2020-06-22)
* \* Attention! (MODX)Evolution.libraries.ddTools >= 0.40.1 is required (not tested in older versions).
* \* Composer.json:
	* \+ `homepage`
	* \+ `authors`.
	* \* `require` → `dd/evolutioncms-libraries-ddtools`: Renamed from `dd/modxevo-library-ddtools`.


## Version 2.2.1 (2020-05-02)
* \* The snippet result will be returned in anyway (empty string for empty result).
* \+ Composer.json.
* \+ README.
* \+ README_ru.
* \+ CHANGELOG.
* \+ CHANGELOG_ru.
* \* Refactoring, small changes, code style.


## Version 2.2 (2017-03-01)
* \* Attention! (MODX)EvolutionCMS.libraries.ddTools >= 0.18 is required.
* \+ Added [JSON](https://en.wikipedia.org/wiki/JSON) support for the `placeholders` parameter.


## Version 2.1 (2016-12-29)
* \* Attention! PHP >= 5.4 is required.
* \* Attention! (MODX)EvolutionCMS >= 1.1 is required.
* \* Attention! (MODX)EvolutionCMS.libraries.ddTools >= 0.16.2 is required.
* \+ Added support of the `@CODE:` keyword prefix in the `name` parameter.
* \* Additional data has to be passed through the `placeholders` parameter must be set as a [query formated string](https://en.wikipedia.org/wiki/Query_string) (the old format is still supported but deprecated).
* \* The `escaping` parameter was renamed as `escapeResultForJS` (with backward compatibility).


## Version 2.0 (2015-10-06)
* \* Attention! (MODX)EvolutionCMS.libraries.ddTools >= 0.12 is required.
* \* The `screening` parameter has been renamed as `escaping`.
* \* The default value of the `escaping` parameter is now equals `0`.
* \* Comments, readme and other minor changes.


## Version 1.4 (2013-09-24)
* \+ Parameter `removeEmptyPlaceholders` that allows placeholders that have not values to be deleted from the parsed chunk has been added.
* \* Minor changes.


## Version 1.3 (2012-03-21)
* \* Attention! (MODX)EvolutionCMS.libraries.ddTools >= 0.2 is required.
* \+ Added ability to transfer additional data to parse into chunk (param `placeholders`).
* \+ Added ability to turn off screening special chars (param `screening`).


## Version 1.2 (2010-09-27)
* \* Minor changes.


<link rel="stylesheet" type="text/css" href="https://raw.githack.com/DivanDesign/CSS.ddMarkdown/master/style.min.css" />
<style>ul{list-style:none;}</style>

src/Snippet.php
<?php
namespace ddGetChunk;

class Snippet extends \DDTools\Snippet {
	protected
		$version = '2.4.0',
		
		$params = [
			//Defaults
			'name' => '',
			'placeholders' => [],
			'removeEmptyPlaceholders' => false,
			'escapeResultForJS' => false
		],
		
		$renamedParamsCompliance = [
			'escapeResultForJS' => 'escaping'
		],
		
		$paramsTypes = [
			'placeholders' => 'objectArray',
		]
	;
		
	/**
	 * run
	 * @version 1.0.1 (2023-05-14)
	 * 
	 * @return {string}
	 */
	public function run(){
		$result = '';
		
		if (!empty($this->params->name)){
			//Получаем чанк
			$result = \ddTools::getTpl($this->params->name);
			
			//Парсим
			$result = \ddTools::parseText([
				'text' => $result,
				'data' => $this->params->placeholders,
				//Удаляем пустые плэйсхолдеры, если нужно
				'removeEmptyPlaceholders' => $this->params->removeEmptyPlaceholders
			]);
			
			//Окончательно парсим
			$result = \ddTools::parseSource($result);
			
			//Экранируем сиволы, если нужно
			if ($this->params->escapeResultForJS){
				$result = \ddTools::escapeForJS($result);
			}
		}
		
		return $result;
	}
}

ddGetChunk.php
<?php
/**
 * ddGetChunk
 * @version 2.2 (2017-03-01)
 * 
 * @desc Snippet gets the chunk contents by its name. For example, it useful to get chunks inside js code.
 * 
 * @uses PHP >= 5.4.
 * @uses MODXEvo >= 1.1.
 * @uses MODXEvo.libraries.ddTools >= 0.18.
 * 
 * @param $name {string_chunkName|string} — Chunk name or code via “@CODE:” prefix. @required
 * @param $placeholders {stirng_json|string_queryFormated} — Additional data as JSON (https://en.wikipedia.org/wiki/JSON) or Query string (https://en.wikipedia.org/wiki/Query_string) has to be passed into the chunk. E. g. `{"width": 800, "height": 600}` or `width=800&height=600`. Default: ''.
 * @param $removeEmptyPlaceholders {0|1} — Placeholders which have not values to be replaced by will be deleted from parsed chunk if the parameter equals 1. Default: 0.
 * @param $escapeResultForJS {0|1} — Escaping special chars (for js). Default: 0.
 * 
 * @link http://code.divandesign.biz/modx/ddgetchunk/2.2
 * 
 * @copyright 2009–2017 DivanDesign {@link http://www.DivanDesign.biz }
 */

if (!empty($name)){
	//Подключаем modx.ddTools
	require_once $modx->getConfig('base_path').'assets/libs/ddTools/modx.ddtools.class.php';
	
	//Для обратной совместимости
	extract(ddTools::verifyRenamedParams(
		$params,
		[
			'escapeResultForJS' => 'escaping'
		]
	));
	
	//Получаем чанк
	$result = $modx->getTpl($name);
	
	//Если переданы дополнительные данные
	if (!empty($placeholders)){
		$placeholders = ddTools::encodedStringToArray($placeholders);
	}else{
		$placeholders = [];
	}
	
	//Парсим
	$result = ddTools::parseText([
		'text' => $result,
		'data' => $placeholders,
		//Удаляем пустые плэйсхолдеры, если нужно
		'removeEmptyPlaceholders' => isset($removeEmptyPlaceholders) &&	$removeEmptyPlaceholders == '1' ? true : false
	]);
	
	//Окончательно парсим
	$result = ddTools::parseSource($result);
	
	//Экранируем сиволы, если нужно
	if (
		isset($escapeResultForJS) &&
		$escapeResultForJS == '1'
	){
		$result = ddTools::escapeForJS($result);
	}
	
	return $result;
}
?>